sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java

```diff
@@ -492,6 +492,7 @@ public void write(final ChannelHandlerContext context, final Object message, fin
 
             final RntbdRequestRecord record = (RntbdRequestRecord) message;
             this.timestamps.channelWriteAttempted();
+            record.setSendingRequestHasStarted();
 
```

Comment:
If the STAGE status is before SENT, is that enough to decide whether we should retry? 

Or maybe can use @flush() to decide whether the request is on wire? I think when calling flush(), sslHandler will start to prepare handshake etc

Reply:
No - Especially the SSL handhsake has relatively high potential for error/timeout - and at that point we would still be "safe to retry" - the right time is really just before we start sending the actual request on the wire. Does that make sense?

Reply:
Hmm, but I thought instead of creating a new flag, we probably should be safe enough to use Stage.Sent to guard the retry?

But this change makes sense to me~

Reply:
Stage.SENT is only set after finishing sending the request on the wire - which means we could receive an error we map to 410 while sending the request - which isn't 100% safe from the idempotency perspective. Once we start sending the request we can't exclude the risk of side effects on the service (because retrieving ACK from the service and setting Stage.SENT isn't happening atomically - although I admit the risk is low. I personally don't think the additional flag is increasing overall complexity much - but as the author I am obviously biased. If there is broader concern about the complexity introduced by this additional flag I am ok with trading-in some correctness and using Stage < SENT as indication of allowing retries. @moderakh / @kirankumarkolli  / @kushagraThapar / @simplynaveen20 - Thoughts?

Reply:
So currently, we are using write() not writeAndFlush(), that is why I think maybe it is safe (since the write method is only queue the message, but not really writing the message to the underlying socket).

Reply:
@FabianMeiswinkel  - I am fine with adding the flag, I don't think it introduces additional complexity.